url9ba68db77a_samkio_Dungeonator_tgz-pJ8-net_woopa_dungeon_core_Main2J8/src/net/woopa/dungeon/datatypes/Grid.java

```diff
@@ -0,0 +1,182 @@
+	}
+
+	public float percentUtilized() {
+		return ((float) used * 100) / ((float) size.getX() * size.getY());
+	}
+
+	public void show() {
+		for (String s : getMap()) {
+			System.out.println(s);
+		}
+	}
+
+	public List<String> getMap() {
+		List<String> list = new ArrayList<String>();
+		String line = "";
+		for (int x = 0; x < size.getX() + 2; x++) {
+			line += CoreMaterial.NONE.getSymbol();
+		}
+		list.add(line);
+		for (int y = size.getY() - 1; y >= 0; y--) {
+			line = "";
+			line += CoreMaterial.NONE.getSymbol();
+			for (int x = 0; x < size.getX(); x++) {
+				Material b = area[x][y];
+				line += b.getSymbol();
+			}
+			line += CoreMaterial.NONE.getSymbol();
+			list.add(line);
+
+		}
+		line = "";
+		for (int x = 0; x < size.getX() + 2; x++) {
+			line += CoreMaterial.NONE.getSymbol();
+		}
+		list.add(line);
+		return list;
+	}
+
+	public Vector2D getSize() {
+		return this.size;
+	}
+
+	public void renderBasicEmptyRoom(int origin_x, int origin_y, int size_x,
+			int size_y, Material wall, Material floor) {
+		if (origin_x + size_x - 1 >= size.getX() || origin_x < 0
+				|| origin_y + size_y - 1 >= size.getY() || origin_y < 0)
+			return;
+
+		for (int x = origin_x; x < origin_x + size_x; x++) {
+			for (int y = origin_y; y < origin_y + size_y; y++) {
+				if (x == origin_x || y == origin_y
+						|| x == origin_x + size_x - 1
+						|| y == origin_y + size_y - 1) {
+					if (isUndef(x, y)) {
+						set(x, y, wall);
+						used++;
+					}
+				} else {
+					if (isUndef(x, y)) {
+						set(x, y, floor);
+						used++;
+					}
+				}
+			}
+		}
+	}
+
+	public Boolean isUndef(int x, int y) {
+		if (x < 0 || y < 0 || x >= size.getX() || y >= size.getY())
+			return false;
+		return (area[x][y]).isUndef();
+	}
+
+	public void set(int x, int y, Material b) {
+		if (x >= 0 && y >= 0 || x < size.getX() || y < size.getY())
+			area[x][y] = b;
+	}
+
+	public Boolean fits(int origin_x, int origin_y, int size_x, int size_y) {
+		if (origin_x + size_x - 1 >= size.getX() || origin_x < 0
+				|| origin_y + size_y - 1 >= size.getY() || origin_y < 0)
+			return false;
+
+		for (int x = origin_x; x < origin_x + size_x; x++) {
+			for (int y = origin_y; y < origin_y + size_y; y++) {
+				if (x == origin_x || y == origin_y
+						|| x == origin_x + size_x - 1
+						|| y == origin_y + size_y - 1) {
+					if (!(area[x][y]).isBoundary() && (!(area[x][y]).isUndef()))
+						return false;
+				} else {
+					if (!area[x][y].isUndef())
+						return false;
+				}
+			}
+		}
+		return true;
+	}
+
+	public Material get(int x, int y) {
+		if (x < 0 || y < 0 || x >= size.getX() || y >= size.getY())
+			return CoreMaterial.FIXEDWALL;
+		return area[x][y];
+	}
+
+	public Boolean isFloor(int x, int y) {
+		if (x < 0 || y < 0 || x >= size.getX() || y >= size.getY())
+			return false;
+		return (area[x][y]).isFloor();
+	}
+
+	public void use() {
+		used++;
+	}
+
+	public void renderSchematic(int origin_x, int origin_y, Schematic map,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.